README.md
# MagicMirror² Module: Ruter
'MMM-Ruter' is a module for displaying public transport information for the Oslo and Akershus region in Norway on a [MagicMirror²](https://magicmirror.builders/). It's using data from [Ruter.no](http://reisapi.ruter.no/help).

![Simple](images/MMM-Ruter_Simple.png)
![Full](images/MMM-Ruter_Full.png) 

Current version is 1.3.0 See [changelog](CHANGELOG.md "Version history") for version history.

## Installation

Remote to your MM2-box with your terminal software and go to your MagicMirror's Module folder:
````bash
cd ~/MagicMirror/modules
````

Clone the repository:
````bash
git clone https://github.com/CatoAntonsen/MMM-Ruter.git
````

Go to the modules folder:
````bash
cd MMM-Ruter
````

Install the dependencies:
````bash
npm install
````

Add the module to the modules array in the `config/config.js` file by adding the following section. You can change this configuration later when you see this works:
```
{
	module: 'MMM-Ruter',
	header: 'Ruter',
	position: 'top_left',
	config: {
		showPlatform: true, 
		maxItems: 10,
		stops: [
			{
				stopId: "3010972",
				platforms: [ "1", "2" ],
				timeToThere: 0
			}
		]
	}
},
```

# Configuration options

These are the valid configuration options you can put inside the config array above:

Configuration option | Comment | Default 
---|---|---
stops | Array of stops. See below | Empty array
maxItems | Number of journeys to display | 5 
showHeader | Set this to true to show header above the journeys | false
showStopName | Display custom stop name for each stop. You can override the name by adding `stopName` to the stop you want to override (See below) | false
maxNameLength | Some stop names can be very long and ruin the layout of the module. Set this to how many characters you max want to display.  | Not set
showPlatform | Set this to true to get the names of the platforms. Set this to true to check the name of the platform if you need to filter  | false
humanizeTimeTreshold | If time to next journey is below this value, it will be displayed as "x minutes" instead of time | 15 
serviceReloadInterval | Refresh rate in MS for how often we call Ruter's web service. NB! Don't set it too low! | 30000 
timeReloadInterval | Refresh rate how often we check if we need to update the time in the GUI | 1000 
animationSpeed | How fast the animation changes when updating mirror - in milliseconds | 0  
fade | Set this to true to fade list from light to dark | true  
fadePoint | Start fading on 1/4th of the list | 0.25

## Stops
You have to configure at least one stop. The module is using the same stop ID's as Ruter does in it's API. The simplest way to find those are to go to [https://ruter.no](https://ruter.no) in your browser and do a search _from_ your stop to somewhere/anywhere. Then extract `NNNNNN` from the resulting URL: https://ruter.no/reiseplanlegger/Mellom/Fra/(NNNNNN)*. 

Notice that you can only use stops, not addresses or areas. You can test if it is a valid stop by enter the following URL in your browser http://reisapi.ruter.no/StopVisit/GetDepartures/NNNNNN (replace NNNNNN with your id).

Stop option | Comment 
---|---
stopId | Id of stop  
stopName | Override name of the stop if you know it by another name or want to keep it short. You have to enable `showStopName` in module configuration. 
platformFilter | The names of the platforms you want to see. Please temporarely enable `showPlatformName` in the module configuration to get the correct platform names before you configure this option. If these names aren't valid, nothing will be displayed.
timeToThere | How long time in minutes does it take for you to get to this stop? It's no point showing journeys that till go in 1 minute if it takes you 5 minutes to get there...

Example:
```
stops: [
	{
		stopId: "3010972",
		platforms: [ "1", "2" ]
	},
	{
		stopId: "3010910",
		timeToThere: 15,
		stopName: "Mitt stopp"
	}
]

``` 
## Translations

This modules is translated to the following languages:

Language | Responsible
---|---
nb (Norwegian) | Me
en (English) | Me

If you add other languages, please make a PR or drop me a line!

# Future enhanchements

1. Make it possible to have multiple instances of this module simultaneously
2. Show exceptions
3. Add filter for individual lines on one platform


CHANGELOG.md
# MMM-Ruter Change Log

## [1.3.0] - 2017-08-19
- The name of the stop will now automatically be fetched if you set `showStopName`. But you can still override it with the stop config value: `stopName`. See [documentation](README.md "MMM-Ruter Documentation") for more information.

## [1.2.0] - 2017-08-13
- Added the option to display a custom stop name in the list. New module config value: `showStopName` and new stop config value: `stopName`. See [documentation](README.md "MMM-Ruter Documentation") for more information.
- Fixed time format

## [1.1.0] - 2017-04-17

- Now it's possible to add multiple instances of the module
- Added some padding between line number and stop name
- Set default animationSpeed to 0 to prevent "blinking" when module updates often
- Moved all service calls and logic from node_helper.js ("backend") to MMM-Ruter.js ("frontent") to simplify a rather complex logic. My initial goal by putting everything in the backend, was to reduce number of calls to the service. That was not a good design decision...


## [1.0.2] - 2016-11-01

- Added timeToThere config value to stops
- Fixed duplication of journeys if manually refreshed

## [1.0.1] - 2016-10-24

- Fixed stupid this/self-bug in the refresh-code that I managed to introduce just before initial commit

## [1.0.0] - 2016-10-23

- Initial version


node_helper.js
/* Magic Mirror
 * Node Helper: Ruter
 *
 * By Cato Antonsen (https://github.com/CatoAntonsen)
 * MIT Licensed.
 */

var NodeHelper = require("node_helper");
var http = require("http");
var async = require("async");
var crypto = require("crypto");
var moment = require("moment");

function addFieldIfNotExist(object, fieldname) {
    if (!object.hasOwnProperty(fieldname)) {
        object[fieldname] = {};
	};
    return object;
};


module.exports = NodeHelper.create({
	start: function() {
		console.log("Starting module: " + this.name);

		this.config = {}; // config struct for each module
		this.allStops = {}; // list of stops for each module
		this.lastMD5 = {}; // hash of last received arrival list
		this.poller = {}; // poll function for each module
	},

	socketNotificationReceived: function(notification, config) {
		if (notification === "CONFIG" + config.identifier) {
			this.config[config.identifier] = config;
			this.allStops[config.identifier] = [];
			this.lastMD5[config.identifier] = [];
			this.initPolling(config.identifier);
			return;
		}
	},

	initPolling: function(identifier) {
		var self = this;

		this.allStops[identifier]=[];
		for(var i=0; i < this.config[identifier].stops.length; i++) {
			this.allStops[identifier].push(this.config[identifier].stops[i]);
		}

		self.startPolling(identifier);
		
		// Reset interval if defined

		this.poller[identifier] = setInterval(function() {
			self.startPolling(identifier);
		}, this.config[identifier].serviceReloadInterval);
	},
	
	startPolling: function(identifier) {
		var self = this;

		if (this.allStops.hasOwnProperty(identifier)) {

			
			async.map(this.allStops[identifier], this.getStopInfo, function(err, result) {
				var stops = [];
				for(var i=0; i < result.length; i++) {
					stops = stops.concat(result[i]);
				}
				stops.sort(function(a,b) {
					var dateA = new Date(a.time);
					var dateB = new Date(b.time);
					return dateA - dateB;
				});

				stops = stops.slice(0, self.config[identifier].maxItems);

				if (self.hasChanged("stops"+identifier, stops)) {
					console.log("Updating journeys to mirror");
					var packet = {
						stops: stops,
						target_identifier: identifier
					};
					self.sendSocketNotification("RUTER_UPDATE", packet);
				}
			});
		}
	},
	
	hasChanged: function(key, value) {
		var md5sum = crypto.createHash("md5");
		md5sum.update(JSON.stringify(value));
		var md5Hash = md5sum.digest("hex");
		if (md5Hash != this.lastMD5[key]) {
			this.lastMD5[key] = md5Hash;
			return true;
		} else {
			return false;
		}
	},
	
	getStopInfo: function(stopItem, callback) {
		var str = "";

		var responseCallback = function(response) {
			
			response.on("data", function(chunk) {
				str += chunk;
			});
			
			response.on("end", function() {

				var shouldAddPlatform = function(platform, platformFilter) {
					if (platformFilter == null || platformFilter.length == 0) { return true; } // If we don't add any interesting platformFilter, then we asume we'll show all
					for(var i=0; i < platformFilter.length; i++) {
						if (platformFilter[i] === platform) { return true; }
					}
					
					return false;
				};
			
				var stops = JSON.parse(str);

				var allStopItems = new Array();

				for(var j = 0; j < stops.length; j++) {
					var journey = stops[j].MonitoredVehicleJourney;
					
					if (shouldAddPlatform(journey.MonitoredCall.DeparturePlatformName, stopItem.platforms)) {
						var numBlockParts = null;
						if (journey.TrainBlockPart != null) {
							numBlockParts = journey.TrainBlockPart.NumberOfBlockParts;
						}
						allStopItems.push({
							stopId: stopItem.stopId,
							lineName: journey.PublishedLineName,
							destinationName: journey.DestinationName,
							time: journey.MonitoredCall.ExpectedDepartureTime,
							platform: journey.MonitoredCall.DeparturePlatformName
						});
					}
				};

				callback(null, allStopItems);
			});
			
			response.on("error", function(error) {
				console.error("------------->" + error)
			});
		}

		var dateParam = ""
		if (stopItem.timeToThere) {
			console.log("Looking for journey in the funture: " + stopItem.timeToThere);
			var min = stopItem.timeToThere;
			dateParam = "?datetime=" + moment(moment.now()).add(min, "minute").format().substring(0, 16);
		} else {
			console.log("Looking for journey right now");
		}
		
		var options = {
			host: "reisapi.ruter.no",
			path: "/StopVisit/GetDepartures/" + stopItem.stopId + dateParam
		};
	
		http.request(options, responseCallback).end();	
	}
});


MMM-Ruter.js
/* Magic Mirror
 * Module: Ruter
 *
 * By Cato Antonsen (https://github.com/CatoAntonsen)
 * MIT Licensed.
 */
 
Module.register("MMM-Ruter",{

	// Default module config.
	defaults: {
		timeFormat: null,			// This is set automatically based on global config
		showHeader: false, 			// Set this to true to show header above the journeys (default is false)
		showPlatform: false,			// Set this to true to get the names of the platforms (default is false)
		showStopName: false,			// Show the name of the stop (you have to configure 'name' for each stop)
		maxItems: 5,				// Number of journeys to display (default is 5)
		humanizeTimeTreshold: 15, 		// If time to next journey is below this value, it will be displayed as "x minutes" instead of time (default is 15 minutes)
		serviceReloadInterval: 30000, 		// Refresh rate in MS for how often we call Ruter's web service. NB! Don't set it too low! (default is 30 seconds)
		timeReloadInterval: 1000, 		// Refresh rate how often we check if we need to update the time shown on the mirror (default is every second)
		animationSpeed: 0,			// How fast the animation changes when updating mirror (default is 0 second)
		fade: true,				// Set this to true to fade list from light to dark. (default is true)
		fadePoint: 0.25				// Start on 1/4th of the list. 
	},

	getStyles: function () {
		return ["ruter.css"];
	},

	getScripts: function() {
		return ["moment.js"];
	},

	getTranslations: function() {
		return {
			en: "translations/en.json",
			nb: "translations/nb.json"
		}
	},

	start: function() {
		console.log(this.translate("STARTINGMODULE") + ": " + this.name);

		this.journeys = [];
		this.previousJourneys = [];
		var self = this;

 		// Set locale and time format based on global config
		moment.locale(config.language);
		if (config.timeFormat === 24) {
				this.config.timeFormat = 'HH:mm';
		} else {
				this.config.timeFormat = 'h:mm A';
		}

		// Just do an initial poll. Otherwise we have to wait for the serviceReloadInterval
		self.startPolling(); 

		setInterval(function() {
			self.startPolling();
		}, this.config.serviceReloadInterval);
		
		setInterval(function() {
			self.updateDomIfNeeded();
		}, this.config.timeReloadInterval);
	},
	
	getDom: function() {
		if (this.journeys.length > 0) {
			
			var table = document.createElement("table");
			table.className = "ruter small";
			
			if (this.config.showHeader) {
				table.appendChild(this.getTableHeaderRow());
			}
			
			for(var i = 0; i < this.journeys.length; i++) {

				var journey = this.journeys[i];
				var tr = this.getTableRow(journey);
				
				// Create fade effect. <-- stolen from default "calendar" module
				if (this.config.fade && this.config.fadePoint < 1) {
					if (this.config.fadePoint < 0) {
						this.config.fadePoint = 0;
					}
					var startingPoint = this.journeys.length * this.config.fadePoint;
					var steps = this.journeys.length - startingPoint;
					if (i >= startingPoint) {
						var currentStep = i - startingPoint;
						tr.style.opacity = 1 - (1 / steps * currentStep);
					}
				}
				
				table.appendChild(tr);
			}
			
			return table;
		} else {
			var wrapper = document.createElement("div");
			wrapper.innerHTML = this.translate("LOADING");
			wrapper.className = "small dimmed";
		}

		return wrapper;
	},

	startPolling: function() {
		var self = this;

		var promises = [];
		for(var i=0; i < this.config.stops.length; i++) {
			promises.push(new Promise((resolv) => {
				this.getStopInfo(this.config.stops[i], function(err, result) {
					resolv(result);
				});
			}));
		}
		
		Promise.all(promises).then(function(promiseResults) {
			if (promiseResults.length > 0) {
				var allJourneys = [];
				for(var i=0; i < promiseResults.length; i++) {
					allJourneys = allJourneys.concat(promiseResults[i])
				}
				
				allJourneys.sort(function(a,b) {
					var dateA = new Date(a.time);
					var dateB = new Date(b.time);
					return dateA - dateB;
				});

				self.journeys = allJourneys.slice(0, self.config.maxItems);
			}
		});
	},
	
	updateDomIfNeeded: function() {
		var needUpdate = false;
		
		for(var i=0; i < this.journeys.length; i++)  {
			var time = this.formatTime(this.journeys[i].time);	
			if (this.previousJourneys[i] == undefined || this.previousJourneys[i].lineName != this.journeys[i].lineName || this.previousJourneys[i].time != time) {
				needUpdate = true;
				this.previousJourneys[i] = {};
				this.previousJourneys[i].lineName = this.journeys[i].lineName;
				this.previousJourneys[i].time = time;
			}
		}
		
		if (needUpdate) {
			this.updateDom(this.config.animationSpeed);
		}
	},

	getStopInfo: function(stopItem, callback) {
		var self = this;

		var HttpClient = function() {
			this.get = function(requestUrl, requestCallback) {
				var httpRequest = new XMLHttpRequest();
				httpRequest.onreadystatechange = function() { 
					if (httpRequest.readyState == 4 && httpRequest.status == 200)
						requestCallback(httpRequest.responseText);
				}

				httpRequest.open( "GET", requestUrl, true );            
				httpRequest.send( null );
			}
		}

		var shouldAddPlatform = function(platform, platformFilter) {
			if (platformFilter == null || platformFilter.length == 0) { return true; } // If we don't add any interesting platformFilter, then we asume we'll show all
			for(var i=0; i < platformFilter.length; i++) {
				if (platformFilter[i] === platform) { return true; }
			}
			
			return false;
		};

		var departureUrl = function() {
			var dateParam = ""
			if (stopItem.timeToThere) {
				var min = stopItem.timeToThere;
				var timeAhead = moment(moment.now()).add(min, "minute").format().substring(0, 16);
				console.log("Looking for journeys " + min + " minutes ahead in time.");
				dateParam = "?datetime=" + timeAhead;
			} else {
				console.log("Looking for current journeys");
			}
			
			return "http://reisapi.ruter.no/StopVisit/GetDepartures/" + stopItem.stopId + dateParam;
		};

		var stopUrl = function() {
			return "http://reisapi.ruter.no/Place/GetStop/" + stopItem.stopId;
		};

		var client = new HttpClient();

		client.get(stopUrl(), function(stopResponse) {
			var stop = JSON.parse(stopResponse);

			client.get(departureUrl(), function(response) {
				var stops = JSON.parse(response);
	
				var allStopItems = new Array();
	
				for(var j = 0; j < stops.length; j++) {
					var journey = stops[j].MonitoredVehicleJourney;
					
					if (shouldAddPlatform(journey.MonitoredCall.DeparturePlatformName, stopItem.platforms)) {
						var numBlockParts = null;
						if (journey.TrainBlockPart != null) {
							numBlockParts = journey.TrainBlockPart.NumberOfBlockParts;
						}

						var stopName = stopItem.stopName ? stopItem.stopName : stop.Name;
						if (self.config.maxNameLength) {
							stopName = stopName.substring(0, self.config.maxNameLength);
						}

						allStopItems.push({
							stopId: stopItem.stopId,
							stopName: stopName,
							lineName: journey.PublishedLineName,
							destinationName: journey.DestinationName,
							time: journey.MonitoredCall.ExpectedDepartureTime,
							platform: journey.MonitoredCall.DeparturePlatformName
						});
					}
				};
	
				callback(null, allStopItems)		
			});
		})
	},
	
	getTableHeaderRow: function() {
		var thLine = document.createElement("th");
		thLine.className = "light";
		thLine.appendChild(document.createTextNode(this.translate("LINEHEADER")));

		var thDestination = document.createElement("th");
		thDestination.className = "light";
		thDestination.appendChild(document.createTextNode(this.translate("DESTINATIONHEADER")));
		
		var thPlatform = document.createElement("th");
		thPlatform.className = "light";
		thPlatform.appendChild(document.createTextNode(this.translate("PLATFORMHEADER")));

		var thStopName = document.createElement("th");
		thStopName.className = "light"
		thStopName.appendChild(document.createTextNode(this.translate("STOPNAMEHEADER")));
		
		var thTime = document.createElement("th");
		thTime.className = "light time"
		thTime.appendChild(document.createTextNode(this.translate("TIMEHEADER")));

		var thead = document.createElement("thead");
		thead.addClass = "xsmall dimmed";
		thead.appendChild(thLine);
		thead.appendChild(thDestination);
		if (this.config.showStopName) { thead.appendChild(thStopName); }
		if (this.config.showPlatform) { thead.appendChild(thPlatform); }
		thead.appendChild(thTime);
		
		return thead;
	},
	
	getTableRow: function(journey) {
		var tdLine = document.createElement("td");
		tdLine.className = "line";
		var txtLine = document.createTextNode(journey.lineName);
		tdLine.appendChild(txtLine);
		
		var tdDestination = document.createElement("td");
		tdDestination.className = "destination bright";
		tdDestination.appendChild(document.createTextNode(journey.destinationName));
		
		if (this.config.showPlatform) {
			var tdPlatform = document.createElement("td");
			tdPlatform.className = "platform";
			tdPlatform.appendChild(document.createTextNode(journey.platform));
		}

		if (this.config.showStopName) {
			var tdStopName = document.createElement("td");
			tdStopName.className = "light";
			tdStopName.appendChild(document.createTextNode(journey.stopName));	
		}
		
		var tdTime = document.createElement("td");
		tdTime.className = "time light";
		tdTime.appendChild(document.createTextNode(this.formatTime(journey.time)));
		
		var tr = document.createElement("tr");
		tr.appendChild(tdLine);
		tr.appendChild(tdDestination);
		if (this.config.showStopName) { tr.appendChild(tdStopName); }
		if (this.config.showPlatform) { tr.appendChild(tdPlatform); }
		tr.appendChild(tdTime);
		
		return tr;
	},
	
	formatTime: function(t) {
		var diff = moment.duration(moment(t) - moment.now());
		var min = diff.minutes() + diff.hours() * 60;

		if (min == 0) {
			return this.translate("NOW")
		} else if (min == 1) {
			return this.translate("1MIN");
		} else if (min < this.config.humanizeTimeTreshold) {
			return min + " " + this.translate("MINUTES");
		} else {
			return moment(t).format(this.config.timeFormat);
		}
	}
});
